### PR TITLE
feat: add firstinstance handling

### PR DESCRIPTION
**Pull request by stebenz**

### CODE REVIEW

File: charts/zitadel/templates/NOTES.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Thank you for installing {{ .Chart.Name }}.
+
+{{- if and (not (empty .Values.zitadel.configmapConfig.FirstInstance)) (and (not (empty .Values.zitadel.configmapConfig.FirstInstance.Org)) (not (empty .Values.zitadel.configmapConfig.FirstInstance.Org.Machine))) }}
```
Comment (eliobischof) [inline charts/zitadel/templates/NOTES.txt:3]: does the following not work?
{{- if and .Values.zitadel.configmapConfig.FirstInstance .Values.zitadel.configmapConfig.FirstInstance.Org .Values.zitadel.configmapConfig.FirstInstance.Org.Machine }}

File: charts/zitadel/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+If this is you first time deploying ZITADEL, you can now extract your service account key with:
```
Comment (eliobischof) [inline charts/zitadel/templates/NOTES.txt:4]: The secret should still be there after `helm upgrade`, right?
```suggestion
Extract your service account key for authenticating at the ZITADEL API's:
```

File: charts/zitadel/templates/rbac.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-delete-policy: before-hook-creation
+    helm.sh/hook-weight: "0"
+rules:
+  - apiGroups: [ "" ]
+    resources: [ "secrets" ]
+    verbs: [ "get", "create" ]
+  - apiGroups: [ "" ]
+    resources: [ "pods" ]
+    verbs: [ "get", "list", "watch" ]
```
Comment (eliobischof) [inline charts/zitadel/templates/rbac.yaml:18]: Don't we just need get?

File: charts/zitadel/templates/setupjob.yaml [hunk 4]
```diff
@@ -79,8 +79,25 @@ spec:
...
             mountPath: /.secrets
+          {{- if and (not (empty .Values.zitadel.configmapConfig.FirstInstance)) (and (not (empty .Values.zitadel.configmapConfig.FirstInstance.Org)) (not (empty .Values.zitadel.configmapConfig.FirstInstance.Org.Machine))) }}
+          - name: machinekey
+            mountPath: {{ dir .Values.zitadel.configmapConfig.FirstInstance.MachineKeyPath }}
+          {{- end}}
           resources:
             {{- toYaml .Values.setupJob.resources | nindent 14 }}
+        {{- if and (not (empty .Values.zitadel.configmapConfig.FirstInstance)) (and (not (empty .Values.zitadel.configmapConfig.FirstInstance.Org)) (not (empty .Values.zitadel.configmapConfig.FirstInstance.Org.Machine))) }}
+        - name: "{{ .Chart.Name}}-machinekey"
+          image: alpine/k8s:1.25.2
```
Comment (eliobischof) [inline charts/zitadel/templates/setupjob.yaml:90]: can we use the kubeapi's k8s version?

File: charts/zitadel/templates/setupjob.yaml [hunk 5]
```diff
@@ -79,8 +79,25 @@ spec:
...
+          command: [ "sh","-c","until [ ! -z $(kubectl get po ${POD_NAME} -o jsonpath=\"{.status.containerStatuses[?(@.name=='{{ .Chart.Name }}-setup')].state.terminated}\") ]; echo 'waiting for {{ .Chart.Name }}-setup container to terminate'; do sleep 5; done && echo '{{ .Chart.Name }}-setup container terminated' && if [ -f {{ .Values.zitadel.configmapConfig.FirstInstance.MachineKeyPath }} ]; then kubectl -n {{ .Release.Namespace }} create secret generic {{ .Values.zitadel.configmapConfig.FirstInstance.Org.Machine.Machine.Username }} --from-file={{ .Values.zitadel.configmapConfig.FirstInstance.Org.Machine.Machine.Username }}.json={{ .Values.zitadel.configmapConfig.FirstInstance.MachineKeyPath }}; fi;" ]
```
Comment (eliobischof) [inline charts/zitadel/templates/setupjob.yaml:91]: This also works if the setup doesn't create a file, you said?

```suggestion
          command: [ "sh","-c","until [ ! -z $(kubectl -n {{ .Release.Namespace }} get po ${POD_NAME} -o jsonpath=\"{.status.containerStatuses[?(@.name=='{{ .Chart.Name }}-setup')].state.terminated}\") ]; echo 'waiting for {{ .Chart.Name }}-setup container to terminate'; do sleep 5; done && echo '{{ .Chart.Name }}-setup container terminated' && if [ -f {{ .Values.zitadel.configmapConfig.FirstInstance.MachineKeyPath }} ]; then kubectl -n {{ .Release.Namespace }} create secret generic {{ .Values.zitadel.configmapConfig.FirstInstance.Org.Machine.Machine.Username }} --from-file={{ .Values.zitadel.configmapConfig.FirstInstance.Org.Machine.Machine.Username }}.json={{ .Values.zitadel.configmapConfig.FirstInstance.MachineKeyPath }}; fi;" ]
```

**Comment by eliobischof** [comment]
Are the role and rolebinding removed with helm uninstall? If not, can you add the commands for manual deletion to the readme, plz?

File: charts/zitadel/templates/setupjob.yaml (hunk 4 above)
Comment (stebenz) [reply charts/zitadel/templates/setupjob.yaml:90]: is the information what k8s-version is used already in the chart?

**Comment by stebenz** [comment]
> Are the role and rolebinding removed with helm uninstall? If not, can you add the commands for manual deletion to the readme, plz?

Seems like a lot of resources are not removed when you use "helm uninstall" is this intended?

File: charts/zitadel/templates/setupjob.yaml (hunk 4 above)
Comment (stebenz) [reply charts/zitadel/templates/setupjob.yaml]: added through build-in helm object "Capabilities.KubeVersion"

**Comment by eliobischof** [comment]
> > Are the role and rolebinding removed with helm uninstall? If not, can you add the commands for manual deletion to the readme, plz?
> 
> Seems like a lot of resources are not removed when you use "helm uninstall" is this intended?

Last time I checked, it just didn't remove the jobs. What other resources did you find after `helm uninstall`?

**Comment by stebenz** [comment]
> Are the role and rolebinding removed with helm uninstall? If not, can you add the commands for manual deletion to the readme, plz?

As it seems none of the resources are deleted with 'helm uninstall my-zitadel'

> > > Are the role and rolebinding removed with helm uninstall? If not, can you add the commands for manual deletion to the readme, plz?
> > 
> > 
> > Seems like a lot of resources are not removed when you use "helm uninstall" is this intended?
> 
> Last time I checked, it just didn't remove the jobs. What other resources did you find after `helm uninstall`?

What I see that is still around after uninstall:
- roles
- rolebindings
- serviceaccount(s)
- configmaps
- secrets

**Comment by eliobischof** [comment]
Linting fails in the pipeline.
Committed the instructions directly to this branch.

File: charts/zitadel/templates/NOTES.txt (hunk 1 above)
Comment (eliobischof) [reply charts/zitadel/templates/NOTES.txt]: ?

**Comment by eliobischof** [comment]
> > Are the role and rolebinding removed with helm uninstall? If not, can you add the commands for manual deletion to the readme, plz?
> 
> As it seems none of the resources are deleted with 'helm uninstall my-zitadel'
> 
> > > > Are the role and rolebinding removed with helm uninstall? If not, can you add the commands for manual deletion to the readme, plz?
> > > 
> > > 
> > > Seems like a lot of resources are not removed when you use "helm uninstall" is this intended?
> > 
> > 
> > Last time I checked, it just didn't remove the jobs. What other resources did you find after `helm uninstall`?
> 
> What I see that is still around after uninstall:
> 
> * roles
> * rolebindings
> * serviceaccount(s)
> * configmaps
> * secrets

You're right. I guess I overlooked them because they are not shown by `kubectl get all`
Created a PR https://github.com/zitadel/zitadel-charts/pull/50

**Comment by eliobischof** [comment]
Didn't you want to add a test that authenticates at ZITADEL using the key from the setup secret?

**Comment by stebenz** [comment]
> Didn't you want to add a test that authenticates at ZITADEL using the key from the setup secret?

Yes, that would be a great idea, will just include one

File: charts/zitadel/templates/NOTES.txt (hunk 1 above)
Comment (stebenz) [reply charts/zitadel/templates/NOTES.txt]: changed it in all ifs

Comment (eliobischof) [reply charts/zitadel/templates/NOTES.txt]: doesn't look changed to me :joy: 
image

Comment (stebenz) [reply charts/zitadel/templates/NOTES.txt]: ah you're right, i just thought that the second 'and' was to much, but the 'not' and 'empty' can be removed too

**Review by eliobischof — APPROVED** [review]

**Pull request merged**